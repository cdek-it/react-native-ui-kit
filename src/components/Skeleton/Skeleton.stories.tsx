import type { Meta, StoryObj } from '@storybook/react'

import { StyleSheet, View } from 'react-native'

import { Skeleton } from './Skeleton'

const styles = StyleSheet.create({
  row: { flexDirection: 'row', gap: 10 },
  rowSkeleton: { height: 50, flex: 1 },
})

const meta: Meta<typeof Skeleton> = {
  title: 'Skeleton',
  component: Skeleton,
  render: () => (
    <View style={{ gap: 10 }}>
      <Skeleton style={{ height: 50 }} />
      <View style={styles.row}>
        <Skeleton style={{ height: 50, width: 100 }} />
        <Skeleton style={styles.rowSkeleton} />
      </View>
      <View style={styles.row}>
        <Skeleton style={styles.rowSkeleton} />
        <Skeleton style={styles.rowSkeleton} />
        <Skeleton style={styles.rowSkeleton} />
        <Skeleton style={styles.rowSkeleton} />
        <Skeleton style={styles.rowSkeleton} />
      </View>
      <View style={styles.row}>
        <Skeleton style={{ width: 50, height: 50, borderRadius: 50 }} />
        <View style={{ gap: 5, justifyContent: 'center' }}>
          <Skeleton style={{ height: 15, width: 100 }} />
          <Skeleton style={{ height: 15, width: 200 }} />
        </View>
      </View>
    </View>
  ),
}

export default meta

type Story = StoryObj<typeof Skeleton>

const SkeletonStory: Story = {}

export { SkeletonStory as Skeleton }
