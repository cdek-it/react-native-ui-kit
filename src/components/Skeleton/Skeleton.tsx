import { memo, useEffect } from 'react'
import { View } from 'react-native'
import Animated, {
  Easing,
  interpolate,
  useAnimatedStyle,
  useSharedValue,
  withRepeat,
  withTiming,
} from 'react-native-reanimated'
import Svg, { Defs, LinearGradient, Rect, Stop } from 'react-native-svg'
import type { ViewProps } from 'react-native-svg/lib/typescript/fabric/utils'

import { makeStyles } from '../../utils/makeStyles'

interface SkeletonProps extends ViewProps {}

export const ANIMATION_DURATION = 1200 // ms

/**
 * Используется для отображения контента в момент загрузки
 * @see https://www.figma.com/design/4TYeki0MDLhfPGJstbIicf/UI-kit-PrimeFace-(DS)?node-id=5241-3731
 */
export const Skeleton = memo<SkeletonProps>(({ style, testID, ...rest }) => {
  const styles = useStyles()
  const animation = useSharedValue(0)
  const animatedStyles = useAnimatedStyle(() => ({
    left: `${interpolate(animation.value, [0, 1], [-100, 100])}%`,
  }))

  useEffect(() => {
    animation.value = withRepeat(
      withTiming(1, { duration: ANIMATION_DURATION, easing: Easing.ease }),
      -1
    )
  }, [animation])

  return (
    <View
      {...rest}
      style={[styles.container, style]}
      testID={testID ?? SkeletonTestId.root}
    >
      <Animated.View
        style={[styles.gradientContainer, animatedStyles]}
        testID={SkeletonTestId.animatedView}
      >
        <Svg testID={SkeletonTestId.svg}>
          <Defs>
            <LinearGradient id='gradient' x1='0' x2='1' y1='1' y2='1'>
              <Stop
                offset='0'
                stopColor={styles.container.backgroundColor}
                stopOpacity='0.4'
              />
              <Stop
                offset='0.5'
                stopColor={styles.gradientColor.backgroundColor}
                stopOpacity='0.4'
              />
              <Stop
                offset='1'
                stopColor={styles.container.backgroundColor}
                stopOpacity='0.4'
              />
            </LinearGradient>
          </Defs>
          <Rect fill='url(#gradient)' height='100%' width='100%' />
        </Svg>
      </Animated.View>
    </View>
  )
})

const useStyles = makeStyles(({ border, theme }) => ({
  container: {
    borderRadius: border.Radius['rounded-lg'],
    overflow: 'hidden',
    backgroundColor: theme.Misc.Skeleton.skeletonBg,
  },
  gradientContainer: {
    position: 'absolute',
    width: '100%',
    height: '100%',
    left: '-100%',
  },
  gradientColor: { backgroundColor: '#FBFBFB' },
}))

export const SkeletonTestId = {
  root: 'Skeleton',
  animatedView: 'Skeleton.animatedView',
  svg: 'Skeleton.svg',
}
