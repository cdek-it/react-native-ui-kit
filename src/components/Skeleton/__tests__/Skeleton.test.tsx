import { render } from '@testing-library/react-native'

import { ANIMATION_DURATION, Skeleton } from '../Skeleton'

describe('Skeleton', () => {
  beforeEach(() => {
    jest.useFakeTimers() // Включаем фейковые таймеры перед каждым тестом
  })

  afterEach(() => {
    jest.runOnlyPendingTimers()
    jest.useRealTimers() // Возвращаем реальное время после тестов
  })

  test('before animation started', () => {
    const { toJSON } = render(<Skeleton style={{ height: 50 }} />)

    expect(toJSON()).toMatchSnapshot('before animation started')
  })

  test('50% animation', () => {
    const { toJSON } = render(<Skeleton style={{ height: 50 }} />)

    jest.advanceTimersByTime(ANIMATION_DURATION / 2)

    expect(toJSON()).toMatchSnapshot()
  })

  test('100% animation', () => {
    const { toJSON } = render(<Skeleton style={{ height: 50 }} />)

    jest.advanceTimersByTime(ANIMATION_DURATION)

    expect(toJSON()).toMatchSnapshot()
  })
})
