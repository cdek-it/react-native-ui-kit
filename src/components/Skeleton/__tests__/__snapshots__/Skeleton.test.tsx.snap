// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skeleton 50% animation 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#dbdada",
        "borderRadius": 7,
        "overflow": "hidden",
      },
      {
        "height": 50,
      },
    ]
  }
  testID="Skeleton"
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "left": "-36.92863748549213%",
        },
      }
    }
    style={
      {
        "height": "100%",
        "left": "-100%",
        "position": "absolute",
        "width": "100%",
      }
    }
    testID="Skeleton.animatedView"
  >
    <RNSVGSvgView
      bbHeight="100%"
      bbWidth="100%"
      focusable={false}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": "100%",
            "width": "100%",
          },
        ]
      }
      testID="Skeleton.svg"
    >
      <RNSVGGroup
        fill={
          {
            "payload": 4278190080,
            "type": 0,
          }
        }
      >
        <RNSVGDefs>
          <RNSVGLinearGradient
            gradient={
              [
                0,
                1725684442,
                0.5,
                1727790075,
                1,
                1725684442,
              ]
            }
            gradientTransform={null}
            gradientUnits={0}
            name="gradient"
            x1="0"
            x2="1"
            y1="1"
            y2="1"
          />
        </RNSVGDefs>
        <RNSVGRect
          fill={
            {
              "brushRef": "gradient",
              "type": 1,
            }
          }
          height="100%"
          propList={
            [
              "fill",
            ]
          }
          width="100%"
          x={0}
          y={0}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</View>
`;

exports[`Skeleton 100% animation 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#dbdada",
        "borderRadius": 7,
        "overflow": "hidden",
      },
      {
        "height": 50,
      },
    ]
  }
  testID="Skeleton"
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "left": "100%",
        },
      }
    }
    style={
      {
        "height": "100%",
        "left": "-100%",
        "position": "absolute",
        "width": "100%",
      }
    }
    testID="Skeleton.animatedView"
  >
    <RNSVGSvgView
      bbHeight="100%"
      bbWidth="100%"
      focusable={false}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": "100%",
            "width": "100%",
          },
        ]
      }
      testID="Skeleton.svg"
    >
      <RNSVGGroup
        fill={
          {
            "payload": 4278190080,
            "type": 0,
          }
        }
      >
        <RNSVGDefs>
          <RNSVGLinearGradient
            gradient={
              [
                0,
                1725684442,
                0.5,
                1727790075,
                1,
                1725684442,
              ]
            }
            gradientTransform={null}
            gradientUnits={0}
            name="gradient"
            x1="0"
            x2="1"
            y1="1"
            y2="1"
          />
        </RNSVGDefs>
        <RNSVGRect
          fill={
            {
              "brushRef": "gradient",
              "type": 1,
            }
          }
          height="100%"
          propList={
            [
              "fill",
            ]
          }
          width="100%"
          x={0}
          y={0}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</View>
`;

exports[`Skeleton before animation started: before animation started 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#dbdada",
        "borderRadius": 7,
        "overflow": "hidden",
      },
      {
        "height": 50,
      },
    ]
  }
  testID="Skeleton"
>
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {
          "left": "-100%",
        },
      }
    }
    style={
      {
        "height": "100%",
        "left": "-100%",
        "position": "absolute",
        "width": "100%",
      }
    }
    testID="Skeleton.animatedView"
  >
    <RNSVGSvgView
      bbHeight="100%"
      bbWidth="100%"
      focusable={false}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": "100%",
            "width": "100%",
          },
        ]
      }
      testID="Skeleton.svg"
    >
      <RNSVGGroup
        fill={
          {
            "payload": 4278190080,
            "type": 0,
          }
        }
      >
        <RNSVGDefs>
          <RNSVGLinearGradient
            gradient={
              [
                0,
                1725684442,
                0.5,
                1727790075,
                1,
                1725684442,
              ]
            }
            gradientTransform={null}
            gradientUnits={0}
            name="gradient"
            x1="0"
            x2="1"
            y1="1"
            y2="1"
          />
        </RNSVGDefs>
        <RNSVGRect
          fill={
            {
              "brushRef": "gradient",
              "type": 1,
            }
          }
          height="100%"
          propList={
            [
              "fill",
            ]
          }
          width="100%"
          x={0}
          y={0}
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
</View>
`;
